GH-35287: [C++][Parquet] Add CodecOptions to customize the compression parameter

yyang52 opened the pull request:
<!--
Thanks for opening a pull request!
If this is your first pull request you can find detailed information on how 
to contribute here:
  * [New Contributor's Guide](https://arrow.apache.org/docs/dev/developers/guide/step_by_step/pr_lifecycle.html#reviews-and-merge-of-the-pull-request)
  * [Contributing Overview](https://arrow.apache.org/docs/dev/developers/overview.html)


If this is not a [minor PR](https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes). Could you open an issue for this pull request on GitHub? https://github.com/apache/arrow/issues/new/choose

Opening GitHub issues ahead of time contributes to the [Openness](http://theapacheway.com/open/#:~:text=Openness%20allows%20new%20users%20the,must%20happen%20in%20the%20open.) of the Apache Arrow project.

Then could you also rename the pull request title in the following format?

    GH-${GITHUB_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

or

    MINOR: [${COMPONENT}] ${SUMMARY}

In the case of PARQUET issues on JIRA the title also supports:

    PARQUET-${JIRA_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

-->

### Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->
Based on https://github.com/apache/arrow/issues/35287, we'd like to add a CodecOptions to make more compression parameters (such as window_bits) customizable when creating the Codec for parquet writer.

Authored-by: Yang Yang [yang10.yang@intel.com](mailto:yang10.yang@intel.com)
Co-authored-by: Rambacher, Mark [mark.rambacher@intel.com](mailto:mark.rambacher@intel.com)

### What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->
Add CodecOptions and replace `compression_level` when creating the Codec. The design is basically based on previous discussions. 
### Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->
Yes
### Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->
Yes, when user creates the `WriterProperties`
<!--
If there are any breaking changes to public APIs, please uncomment the line below and explain which changes are breaking.
-->
<!-- **This PR includes breaking changes to public APIs.** -->

<!--
Please uncomment the line below (and provide explanation) if the changes fix either (a) a security vulnerability, (b) a bug that caused incorrect or invalid data to be produced, or (c) a bug that causes a crash (even when the API contract is upheld). We use this to highlight fixes to issues that may affect users without their knowledge. For this reason, fixing bugs that cause errors don't count, since those are usually obvious.
-->
<!-- **This PR contains a "Critical Fix".** -->

mapleFU reviewed (commented):
Seems parquet part change all compression level to options. Do we need to keep api compatible?

yyang52 commented on cpp/src/parquet/properties.h:
```diff
@@ -172,8 +172,8 @@ class PARQUET_EXPORT ColumnProperties {
     max_stats_size_ = max_stats_size;
   }
 
-  void set_compression_level(int compression_level) {
```
For now I replace the compression_level with CodecOptions, could also keep that for compatibility :)

yyang52 replied:
seems like python codec api also takes compression_level as input, which also rely on cpp API (where we changed to Codec::Create(Compression::type, CodecOptions)). should we also modify python API or keep Codec::Create(Compression::type, compression_level)?

mapleFU replied:
I guess both can be used

mapleFU commented on cpp/src/parquet/column_writer.h:
```diff
@@ -87,8 +88,9 @@ class PARQUET_EXPORT PageWriter {
       std::shared_ptr<ArrowOutputStream> sink, Compression::type codec,
-      int compression_level, ColumnChunkMetaDataBuilder* metadata,
-      int16_t row_group_ordinal = -1, int16_t column_chunk_ordinal = -1,
+      const std::shared_ptr<CodecOptions>& codec_options,
```
This should be put to the end for capability

pitrou replied:
Also, unless the `PageWriter` needs to keep ownership of the codec options, this should simply be `const CodecOptions& = {}`.

wgtmac replied:
In addition to the comments above, should be consolidate `Compression::type codec` and `const std::shared_ptr<CodecOptions>& codec_options`? codec_options should know codec.

mapleFU commented on cpp/src/parquet/properties.h:
```diff
@@ -192,6 +197,8 @@ class PARQUET_EXPORT ColumnProperties {
 
   int compression_level() const { return compression_level_; }
```
Since `Codec` contains `compression_level_`, would this be duplicated or useless?

mapleFU replied:
Now we have a `compression_level` and a `CodecOptions`. Previously, user can set `compression_level`, and the compression level would be changed.

After this patch, if user change the compression level, the codec will not be changed.

Can we make `CodecOptions` mutate it's `compression_level`, and make `set_compression_level` set the level in `CodecOptions` directly?

yyang52 replied:
that makes sense :)

mapleFU replied:
Can we just get it from `codec_options_`?

wgtmac replied:
+1 on this. `compression_level_` is duplicated.

mapleFU commented:
cc @pitrou @wgtmac

wgtmac commented on cpp/src/parquet/column_writer.cc:
```diff
@@ -683,20 +683,21 @@ class BufferedPageWriter : public PageWriter {
   if (buffered_row_group) {
     return std::unique_ptr<PageWriter>(new BufferedPageWriter(
-        std::move(sink), codec, compression_level, metadata, row_group_ordinal,
+        std::move(sink), codec, codec_options, metadata, row_group_ordinal,
```
I am not sure if we are able to move `codec` into `codec_options` as well. I don't have a preference here.

yyang52 commented on cpp/src/arrow/util/compression.h:
```diff
@@ -124,7 +168,9 @@ class ARROW_EXPORT Codec {
   /// \brief Create a codec for the given compression algorithm
   static Result<std::unique_ptr<Codec>> Create(
-      Compression::type codec, int compression_level = kUseDefaultCompressionLevel);
+      Compression::type codec,
```
Yes, that could solve the compatibility issue of different APIs. Just to make it clear, you suggest marking the newly added API with CodecOptions as deprecated right? Or the previous one?

wgtmac replied:
I mean deprecating the existing one.

yyang52 replied:
Great, that makes sense :)

pitrou commented on cpp/src/parquet/column_writer.cc:
```diff
@@ -249,9 +249,9 @@ int LevelEncoder::Encode(int batch_size, const int16_t* levels) {
-                       int compression_level, ColumnChunkMetaDataBuilder* metadata,
-                       int16_t row_group_ordinal, int16_t column_chunk_ordinal,
-                       bool use_page_checksum_verification,
+                       const std::shared_ptr<CodecOptions>& codec_options,
```
No need for `shared_ptr` here.
```suggestion
                       const CodecOptions& codec_options,
```

pitrou commented on cpp/src/parquet/column_writer.h:
```diff
@@ -21,6 +21,7 @@
 #include <cstring>
 #include <memory>
 
+#include "arrow/util/compression.h"
```
```suggestion
#include "arrow/type_fwd.h"
```

pitrou commented on cpp/src/parquet/column_writer.h:
```diff
@@ -87,8 +88,9 @@ class PARQUET_EXPORT PageWriter {
 
   static std::unique_ptr<PageWriter> Open(
       std::shared_ptr<ArrowOutputStream> sink, Compression::type codec,
-      int compression_level, ColumnChunkMetaDataBuilder* metadata,
```
Let's not remove this API but create a new one taking `const CodecOptions&`, to avoid breaking API compatibility.

wgtmac replied:
Should we mark the old one as deprecated? The new one should always be preferable and we don't want to maintain different overloads in the long term.

pitrou replied:
We can indeed.

pitrou commented on cpp/src/parquet/types.cc:
```diff
@@ -51,9 +51,30 @@ bool IsCodecSupported(Compression::type codec) {
+}
+
+// Deprecated and use CodecOptions to create Codec instead
 std::unique_ptr<Codec> GetCodec(Compression::type codec, int compression_level) {
```
Here as well, can you implement this overload in terms of the one above?

pitrou commented on r/src/compression.cpp:
```diff
@@ -23,7 +23,8 @@
                                                          R_xlen_t compression_level) {
-  return ValueOrStop(arrow::util::Codec::Create(codec, compression_level));
+  return ValueOrStop(arrow::util::Codec::Create(
+      codec, std::make_shared<CodecOptions>(CodecOptions(compression_level))));
```
Can probably revert this.

pitrou commented:
@yyang52 There are Python test failures on CI.

Also, is it possible to pass a base `CodecOptions` when a derived class exists? (e.g. for gzip). Ideally it should be supported.

wgtmac commented on cpp/src/arrow/util/compression.h:
```diff
@@ -124,7 +168,15 @@ class ARROW_EXPORT Codec {
+
+  /// \brief Create a codec for the given compression algorithm
+  /// \deprecated and left for backwards compatibility.
+  ARROW_DEPRECATED("Use CodecOptions to create Codec")
```
Sounds reasonable. I was thinking it is always preferable to use the new `CodecOptions` to set the level.

yyang52 replied:
That makes more sense, will keep the existing one as it was

yyang52 commented on cpp/src/arrow/util/compression_zlib.cc:
```diff
@@ -17,15 +17,14 @@
 #include "arrow/util/compression_internal.h"
 
+#include <zconf.h>
+#include <zlib.h>
```
Sure, that might be due to some local format auto-change, will revert those.

yyang52 commented:
> @yyang52 There are Python test failures on CI.
> 
> Also, is it possible to pass a base `CodecOptions` when a derived class exists? (e.g. for gzip). Ideally it should be supported.

Yes, I think it supports that (as checked in TEST(TestCodecMisc, SpecifyCompressionLevel) to pass base CodecOptions to different Codecs)

mapleFU commented on cpp/src/arrow/util/compression.h:
```diff
@@ -107,6 +107,46 @@ class ARROW_EXPORT Decompressor {
+// ----------------------------------------------------------------------
+// brotli codec options implementation
+
+constexpr int kBrotliDefaultWindowBits = 22;
```
Can it use `BROTLI_DEFAULT_WINDOW`?

yyang52 replied:
yes, but then we need to include the `brotli` header here as well?

yyang52 replied:
yes, I think that's just how it is implemented now in `GZip/BrotliCodecOptions`?

mapleFU replied:
Sure, I notice that this patch hard code the encoder's internal into `compression.h`, so I'm afraid if they're changed in the 3rd libraries, we'll get confused here.

yyang52 replied:
yeah it seems better to use their `BROTLI_DEFAULT_WINDOW`

wgtmac replied:
Same with my previous comment. We can use `BROTLI_DEFAULT_WINDOW` in the `.cc` file

mapleFU commented on cpp/src/parquet/column_io_benchmark.cc:
```diff
@@ -40,8 +40,8 @@ std::shared_ptr<Int64Writer> BuildWriter(int64_t output_size,
                                          Compression::type codec) {
-  std::unique_ptr<PageWriter> pager =
-      PageWriter::Open(dst, codec, Codec::UseDefaultCompressionLevel(), metadata);
+    std::unique_ptr<PageWriter> pager =
```
fmt?

mapleFU commented on cpp/src/parquet/column_writer.cc:
```diff
@@ -690,26 +690,40 @@ class BufferedPageWriter : public PageWriter {
   if (buffered_row_group) {
     return std::unique_ptr<PageWriter>(new BufferedPageWriter(
-        std::move(sink), codec, compression_level, metadata, row_group_ordinal,
+        std::move(sink), codec, *codec_options.get(), metadata, row_group_ordinal,
```
I think just `*codec_options` is ok?

mapleFU commented on cpp/src/arrow/util/compression.h:
```diff
@@ -107,6 +107,50 @@ class ARROW_EXPORT Decompressor {
+  virtual ~CodecOptions() = default;
+
+  int compression_level_;
+};
```
I guess @yyang52 's idea is that `CodecOptions` can explicit built from `compression_level`, am I right?

yyang52 replied:
yes, then we could do one-line built for CodecOptions with `compression_level`, which has been used in several places

wgtmac replied:
That sounds reasonable for now. But it would be harder to use once we add more parameters.

pitrou replied:
@yyang52 I'm not sure I understand what you're saying. Both definitions allow writing `CodecOptions{compression_level}`, right?

yyang52 replied:
Per tests, we may need to do this? Am I missing something?
```  
CodecOptions codec_options;
codec_options.compression_level_ = compression_level;
return Codec::Create(codec_type, codec_options);
```

pitrou replied:
Doesn't `CodecOptions{compression_level}` work?

yyang52 replied:
It gives me the compilation error:
```
error: no matching constructor for initialization of 'arrow::util::CodecOptions'
  return Codec::Create(codec_type, CodecOptions{compression_level});
                                   ^           ~~~~~~~~~~~~~~~~~~~
/home/yangyang/arrow/cpp/src/arrow/util/compression.h:111:20: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'int' to 'const arrow::util::CodecOptions' for 1st argument
class ARROW_EXPORT CodecOptions {
                   ^
/home/yangyang/arrow/cpp/src/arrow/util/compression.h:111:20: note: candidate constructor (the implicit default constructor) not viable: requires 0 arguments, but 1 was provided
1 error generated.
```

pitrou replied:
Ah, sorry! Let's forget it then.

mapleFU reviewed (commented):
I'm general ok with the parquet side changes now, there are some comments.

yyang52 commented on cpp/src/arrow/util/compression.h:
```diff
@@ -107,6 +107,46 @@ class ARROW_EXPORT Decompressor {
+// ----------------------------------------------------------------------
+// gzip codec options implementation
+
+struct GZipFormat {
```
I just move `GZipFormat` from `compression_internal.h` to `compression.h` to be able to use it in `CodecOptions` without changing the type.

wgtmac replied:
It was an internal class in the past. As it is moved to a public header now, changing it to enum class would be preferable.

mapleFU commented on cpp/src/parquet/properties.h:
```diff
@@ -174,6 +175,16 @@ class PARQUET_EXPORT ColumnProperties {
 
   void set_compression_level(int compression_level) {
     compression_level_ = compression_level;
```
Can we just remove `compression_level_`?

pitrou replied:
+1, I think we can just keep the codec options.

mapleFU reviewed (commented):
The parquet part looks ok. However:
1. Should we remove "compression_level_` in Builder and properties, and make the interface as a wrapper for visiting `CodecOptions`?
2. Would `CodecOptions` with explicit cast be API broken @pitrou ?

pitrou commented on cpp/src/parquet/column_writer_test.cc:
```diff
@@ -162,6 +163,25 @@ class TestPrimitiveWriter : public PrimitiveTypedTest<TestType> {
+  void TestRequiredWithCodecOptions(
+      Encoding::type encoding, Compression::type compression, bool enable_dictionary,
+      bool enable_statistics, int64_t num_rows = SMALL_SIZE,
+      int compression_level = Codec::UseDefaultCompressionLevel(),
+      ::arrow::util::GZipFormat::type format = ::arrow::util::GZipFormat::GZIP,
+      int window_bits = ::arrow::util::kGZipDefaultWindowBits,
+      bool enable_checksum = false) {
```
Why not simply take a `const CodecOptions&` here?

pitrou commented on cpp/src/parquet/properties.h:
```diff
@@ -154,6 +154,7 @@ class PARQUET_EXPORT ColumnProperties {
         statistics_enabled_(statistics_enabled),
         max_stats_size_(max_stats_size),
         compression_level_(Codec::UseDefaultCompressionLevel()),
+        codec_options_(std::make_shared<CodecOptions>()),
```
Can we just leave this null-initialized by default?

pitrou commented on cpp/src/parquet/properties.h:
```diff
@@ -618,6 +677,7 @@ class PARQUET_EXPORT WriterProperties {
     std::unordered_map<std::string, Encoding::type> encodings_;
     std::unordered_map<std::string, Compression::type> codecs_;
     std::unordered_map<std::string, int32_t> codecs_compression_level_;
```
Let's remove this and only keep `codec_options_`?

pitrou reviewed (commented):
Thanks for the update @yyang52 ! I think this still needs some improvement, see comments below.

wgtmac commented on cpp/src/arrow/util/compression.h:
```diff
@@ -107,6 +107,46 @@ class ARROW_EXPORT Decompressor {
+};
+
+// ----------------------------------------------------------------------
+// gzip codec options implementation
```
```suggestion
// GZip codec options implementation
```

wgtmac commented on cpp/src/arrow/util/compression.h:
```diff
@@ -107,6 +107,46 @@ class ARROW_EXPORT Decompressor {
+class ARROW_EXPORT GZipCodecOptions : public CodecOptions {
+ public:
+  GZipFormat::type gzip_format = GZipFormat::GZIP;
+  int window_bits = kGZipDefaultWindowBits;
```
This could be easily out of sync with the upstream. What about using `std::optional<int>` and always use the library default value if it is unset?

pitrou replied:
That sounds like a good idea.

mapleFU replied:
I also vote for this

wgtmac commented on cpp/src/parquet/column_writer.h:
```diff
@@ -21,6 +21,7 @@
 #include <cstring>
 #include <memory>
 
+#include "arrow/type_fwd.h"
```
Is this required? It should be included by `parquet/platform.h`.

wgtmac commented on cpp/src/parquet/properties.h:
```diff
@@ -190,7 +193,14 @@ class PARQUET_EXPORT ColumnProperties {
+
+  const std::shared_ptr<CodecOptions>& codec_options() const {
+    if (!codec_options_) {
+      return std::make_shared<CodecOptions>();
```
This is a problem:
```
C:\projects\arrow\cpp\src\parquet\properties.h(200) : error C2220: the following warning is treated as an error
C:\projects\arrow\cpp\src\parquet\properties.h(200) : warning C4172: returning address of local variable or temporary
ninja: build stopped: subcommand failed.
```

yyang52 replied:
this seems to be fixed, but there are Python errors that I have no ideas on the hints

yyang52 commented:
Addressed most comments with most CI checks passed (2 fails seem also happening on other PRs). One left thing is hardcoding `window_bits` in `CodecOptions` may easily be out-of-sync with third-party dependency, as mentioned by @wgtmac. Do you have any suggestions/thoughts on that? Please also have a look at other changes I made and see if any other improvements needed. Thanks! @pitrou @mapleFU @wgtmac

wgtmac commented:
https://github.com/apache/arrow/actions/runs/5367297586/jobs/9737382909?pr=35886

There seems to be lint error.

yyang52 commented:
> https://github.com/apache/arrow/actions/runs/5367297586/jobs/9737382909?pr=35886
> 
> There seems to be lint error.

Doing some fixes, could you help to re-trigger other checks? Thanks!

pitrou commented on cpp/src/parquet/platform.h:
```diff
@@ -24,7 +24,8 @@
 #include "arrow/status.h"         // IWYU pragma: export
 #include "arrow/type_fwd.h"       // IWYU pragma: export
-#include "arrow/util/macros.h"    // IWYU pragma: export
+#include "arrow/util/compression.h"
```
I don't think adding this inclusion is necessary, right?

yyang52 replied:
If not, then it will prevent us to initialize as `CodecOptions{}`? Am I doing this right?
```
column_writer.h:101:43: error: invalid use of incomplete type 'parquet::CodecOptions' (aka 'arrow::util::CodecOptions')
      const CodecOptions& codec_options = CodecOptions{});
                                          ^~~~~~~~~~~~~~
```

mapleFU replied:
Maybe forward declr like `default_memory_pool()` can be used?

pitrou replied:
@yyang52 You don't need it in `platform.h`.

yyang52 replied:
yes, could I include that directly in `column_writer.h`, as the forward declaration is not sufficient to initialize an object as far as I learned?

mapleFU commented on cpp/src/arrow/util/compression.h:
```diff
@@ -107,6 +107,44 @@ class ARROW_EXPORT Decompressor {
+  }
+  virtual ~CodecOptions() = default;
+
+  int compression_level_;
```
So would this be changed to a `std::nullopt`?

yyang52 replied:
yeah I made some changes to these part, will push those soon.

yyang52 commented:
May you help to review the code change as well as trigger the remaining CI checks? Thanks! @mapleFU @pitrou @wgtmac

yyang52 commented:
There're 6 failing checks, but seems like not introduced by this PR?

yyang52 commented:
Hi, may I know if there is any change needed or concerns regarding this PR? Thanks! 
@mapleFU @pitrou @wgtmac

mapleFU commented:
I'm a little busy these days, I'll take a look tonight

yyang52 commented:
> I'm a little busy these days, I'll take a look tonight

Sure, take your time. Just let me know if more actions needed, thanks!

pitrou commented:
@yyang52 We're busy preparing a release right now. You might start by merging/rebasing from main to get a more passing CI.

yyang52 commented:
> @yyang52 We're busy preparing a release right now. You might start by merging/rebasing from main to get a more passing CI.

Sure, no rush! I will rebase from main and re-run CI checks.

yyang52 commented:
> May you help to review the code change as well as trigger the remaining CI checks?

May you help to trigger the remaining CI checks again? Thanks! @pitrou

pitrou commented on cpp/src/parquet/column_writer.h:
```diff
@@ -85,6 +87,22 @@ class PARQUET_EXPORT PageWriter {
+      OffsetIndexBuilder* offset_index_builder = NULLPTR,
+      const CodecOptions& codec_options = CodecOptions{});
+
+  ARROW_DEPRECATED("Use the one with CodecOptions instead")
```
```suggestion
  ARROW_DEPRECATED("Deprecated in 13.0.0. Use CodecOptions-taking overload instead.")
```

pitrou commented on cpp/src/arrow/util/compression.h:
```diff
@@ -107,6 +108,40 @@ class ARROW_EXPORT Decompressor {
+  }
+  virtual ~CodecOptions() = default;
+
+  int compression_level_ = kUseDefaultCompressionLevel;
```
Forgot this, but style nit (this is a public member, it's weird to add a trailing underscore):
```suggestion
  int compression_level = kUseDefaultCompressionLevel;
```

pitrou reviewed (changes requested):
@yyang52 There are a couple of comments left to address, can you take a look?

mapleFU reviewed (approved):
Parquet part looks ok to me

yyang52 commented:
> @yyang52 There are a couple of comments left to address, can you take a look?

Doing some fix, may you help to take a look and trigger remaining CI checks? Thanks! @pitrou

wgtmac reviewed (approved):
LGTM. Nice work!

pitrou commented:
@github-actions crossbow submit -g cpp

pitrou commented:
Thanks a lot for your contribution @yyang52 ! I hope you didn't mind the delays.

yyang52 commented:
> 

Thanks for your kind support! It's a pleasure to make contribution to Arrow community! @pitrou @mapleFU @wgtmac

mapleFU commented on cpp/src/parquet/file_writer.cc:
```diff
@@ -291,12 +304,24 @@ class RowGroupSerializer : public RowGroupWriter::Contents {
+      if (!codec_options) {
+        pager = PageWriter::Open(sink_, properties_->compression(path), col_meta,
+                                 row_group_ordinal_, static_cast<int16_t>(column_ordinal),
+                                 properties_->memory_pool(), false, meta_encryptor,
```
This part lost `buffered_row_group_`, and set it to false. Sorry that I didn't found it at first time

yyang52 replied:
Sorry for this missing and thanks for the fix!
